### ISSUE TITLE
Bug: detectAdConfiguration can crash when ticking a setting

### ISSUE THREAD

**Issue by AndroidDeveloperLB** [bocops/UMP-workarounds#2]
I use this to simulate the ad-consent:

```
            val params = ConsentRequestParameters
                .Builder()
                .setTagForUnderAgeOfConsent(false)
            if (BuildConfig.DEBUG) {
                val debugGeography =
                        ConsentDebugSettings.DebugGeography.DEBUG_GEOGRAPHY_EEA  
                val debugSettings = ConsentDebugSettings.Builder(app)
                    .setDebugGeography(debugGeography)
                    .addTestDeviceHashedId(SystemUtils.getDeviceHashedId(app)!!.uppercase())
                    .build()
                params.setConsentDebugSettings(debugSettings)
            }
```

Issue:

When I choose "Manage options", and then turn on the "store and/or access information on a device":

![image](https://github.com/bocops/UMP-workarounds/assets/5357526/ea98b8dd-bdf3-4a97-893a-8fba6baca780)

-it crashes when I check it out on the next time I start the app's process (if it's right after the UI, I get "None"), with the exception of:

```
java.lang.StringIndexOutOfBoundsException: length=1; index=2
   at java.lang.String.charAt(Native Method)
```

For the line of:

```
(tcConsentString[0] == '1' && tcConsentString[2] == '1' && tcConsentString[3] == '1') -> AdConfiguration.ALL
```

Meaning:
 
https://github.com/bocops/UMP-workarounds/blob/main/detect_insufficient_consent/android/kotlin/detect_insufficient_consent.kt#L55

**Comment by AndroidDeveloperLB** [comment]
About the exception, perhaps this simple function could help:

```
    // Check if a binary string has a "1" at position "index" (1-based)
    private fun hasAttribute(input: String, index: Int): Boolean {
        return input.length >= index && input[index - 1] == '1'
    }
```

Source:
https://stackoverflow.com/a/68310602/878126

**Comment by AndroidDeveloperLB** [comment]
Perhaps use `getOrNull` for each character check in the string.

**Comment by bocops** [comment]
Can you please open different issues for unrelated questions? Discussion would get hard to follow if we did all six(?) issues in this thread. Thanks. :)

**Comment by AndroidDeveloperLB** [comment]
Sorry. The main thing here is the bug that causes a crash.

**Comment by AndroidDeveloperLB** [comment]
Perhaps this could help:

https://stackoverflow.com/questions/65351543/how-to-implement-ump-sdk-correctly-for-eu-consent/68310602#68310602
https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/TCFv2/IAB%20Tech%20Lab%20-%20CMP%20API%20v2.md#in-app-details

?

It also removes the need of "goodVendorConfiguration" ...

**Comment by bocops** [comment]
This is fixed with the latest commit.

**Comment by AndroidDeveloperLB** [comment]
Seems you've used getOrNull, but why create a new function for this...
getOrNull is already a function...

Here's my suggestion, including optimization of loading only what's needed based on current situation:

```
/**
 * Checks the stored IABTCF configuration and returns one of the values defined in [AdConfiguration],
 * based on the necessary minimum consent/interest defined here: https://support.google.com/admob/answer/9760862
 */
fun detectAdConfiguration(context: Context): AdConfiguration {
    // default string for "no consent", used in cases where no configuration has previously been stored
    val defaultPurposeString = "0000000000"
    // IABTCF strings are stored in SharedPreferences
    val sharedPrefs = PreferenceUtil.getDefaultSharedPreferences(context)
    // https://developers.google.com/admob/android/privacy/ad-serving-modes
    //limited ads: nothing for 1, consent/legitimate:  2,7,9,10
    //non personalized ads: Consent: 1	, consent/legitimate:  2,7,9,10
    //personalized: consent: 1,3,4 consent/legitimate:  2,7,9,10
    //
    // relevant strings are those for purpose consent or legitimate interest, as well as vendors
    val tcConsentString = sharedPrefs
        .getString("IABTCF_PurposeConsents", defaultPurposeString) ?: defaultPurposeString
    val tcInterestString = sharedPrefs
        .getString("IABTCF_PurposeLegitimateInterests", defaultPurposeString) ?: defaultPurposeString
    //        Log.d("AppLog", "detectAdConfiguration tcConsentString:$tcConsentString tcInterestString:$tcInterestString tcVendorString:$tcVendorString ")
    // in any case we need at least legitimate interest for purposes N = 2, 7, 9 and 10,
    // stored in positions N-1 of either purpose string:
    val sufficientInterest =
            (tcConsentString.getOrNull(1) == '1' || tcInterestString.getOrNull(1) == '1') &&
                    (tcConsentString.getOrNull(6) == '1' || tcInterestString.getOrNull(6) == '1') &&
                    (tcConsentString.getOrNull(8) == '1' || tcInterestString.getOrNull(8) == '1') &&
                    (tcConsentString.getOrNull(9) == '1' || tcInterestString.getOrNull(9) == '1')
    if (!sufficientInterest) {
        return AdConfiguration.NONE
    }
    val tcVendorString = sharedPrefs
        .getString("IABTCF_VendorConsents", "0") ?: "0"
    // TODO vendor configuration is variable, so needs to be defined by the individual developer
    //   - run app and make sure that a valid configuration is stored
    //   - have the app log the value of [tcVendorString], then copy that value to the following line
    //   - repeat if ad configuration changes, perhaps make this value available via remote configuration instead
    val goodVendorConfiguration = "TODO"
    // if the stored string is shorter than what is necessary, at least some vendors will not be
    // configured properly.
    if (tcVendorString.length < goodVendorConfiguration.length) {
        return AdConfiguration.UNCLEAR
    }
    // we need consent for the following purposes N, stored in positions N-1 of the consent string:
    //   1, 3 and 4 to show all ads
    //   1 to show non-personalized ads
    //   no consent to show limited ads
    val maxAdDisplayConfiguration = when {
        tcConsentString.getOrNull(0) != '1' -> AdConfiguration.LIMITED
        tcConsentString.getOrNull(2) == '1' && tcConsentString.getOrNull(3) == '1' -> AdConfiguration.ALL
        else -> AdConfiguration.NONPERSONALIZED
    }
    // build a regex that must match all '1' but not the '0' characters in goodVendorConfiguration,
    // and allows this configuration to be shorter than the string it is compared with
    val vendorRegex = Regex(goodVendorConfiguration.replace("0", ".").plus(".*"))
    //if the regex matches, at least some ads should be served; if not, vendor string is unclear
    return if (vendorRegex.matches(tcVendorString)) {
        maxAdDisplayConfiguration
    }
    else {
        return AdConfiguration.UNCLEAR
    }
}
```
